addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.dsmr.internal.discovery;
+
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.i18n.LocaleProvider;
+import org.eclipse.smarthome.core.i18n.TranslationProvider;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.dsmr.DSMRBindingConstants;
+import org.openhab.binding.dsmr.internal.device.DSMRDeviceRunnable;
+import org.openhab.binding.dsmr.internal.device.DSMREventListener;
+import org.openhab.binding.dsmr.internal.device.DSMRSerialAutoDevice;
+import org.openhab.binding.dsmr.internal.device.connector.DSMRConnectorErrorEvent;
+import org.openhab.binding.dsmr.internal.device.connector.SerialPortManager;
+import org.openhab.binding.dsmr.internal.device.cosem.CosemObject;
+import org.openhab.binding.dsmr.internal.device.p1telegram.P1Telegram;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gnu.io.CommPortIdentifier;
+
+/**
+ * This implements the discovery service for detecting new DSMR Meters.
+ *
+ * The service will iterate over the available serial ports and open the given serial port and wait for telegrams. If
```

Comment:
Hm, we do not do such a port scanning in any other serial binding.
Isn't this mechanism pretty dangerous? I could imagine all kinds of connected serial devices that behave strangely if you open their port and get data from them. I would claim that a binding should not blindly connect to any serially attached device.

IF you claim that it is not dangerous: In such a case, we should introduce a SerialDiscoveryService in the framework and have bindings implement SerialDiscoveryParticipants, similar to UPnP, mDNS and Bluetooth.

In both cases my suggestion would be to remove this discovery service here.

Reply:
> it just waits for data to be received. So I don't think it's dangerous.

Well, some devices might become active in some (unintended) way by the fact that they are connected to and they can start sending data.
But ok, this discovery service is not automatically started (if I see it correctly), but only manually triggered by the user - so the risk is probably minimal (or at least avoidable by not triggering a discovery).

> I'm not sure if that can be generalized in a way that adds much value.

My thinking was that for background discovery, it would definitely make sense to centralize it - e.g. upon startup, the existing serial ports could be scanned for supported devices and also during runtime, if a device is plugged in and a new serial port becomes available, it would be cool to automatically populate the Inbox.
But if such logic were in the individual bindings, they would be concurrently trying to access the ports, which results in conflicts - a central SerialDiscoveryService could instead hand the port to each interested participant one after another. But that would be a topic for a different PR in ESH - if you find it interesting, feel free to create such a PR :-)

Reply:
Yes the discovery process is only manually. That should reduce the risk. In most cases there are very few serial connections I guess (on average 2-3 com ports per system?) . So adding a new serial connected device probably requires some manual handwork to plugin the device in the openHAB server, and thus the use case for an auto discovery service is very low. It's not unlikely the device connected to the serial port does have multiple connections to other devices and there an auto discover might be useful (like rfxcom and this binding). But at that point it's not specific to the serial port anymore, but the protocol of the device and it's then on the port already assigned to the binding.
Also background discovery using the port scanning, might indeed be dangerous, as it does temporary lock the port each scan, and block the port for the actual device. I think as long as it's a manual one time process, while the user is looking should be fine. So I wouldn't recommend to doing background port scanning on a regular basis anyway.